// @kiwicom/orbit-components/lib/Slider/components/Handle/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.isFirst = exports.calculateLeftPosition = void 0;

var React = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _convertHexToRgba = _interopRequireDefault(require("@kiwicom/orbit-design-tokens/lib/convertHexToRgba"));

var _rtl = require("../../../utils/rtl");

var _defaultTheme = _interopRequireDefault(require("../../../defaultTheme"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

var calculateLeftPosition = function calculateLeftPosition(valueNow, valueMin, valueMax, isFirst, isSimple) {
  // if first, stick it to the left edge
  if (isFirst) {
    // if simple (one handle and without histogram)
    if (isSimple) {
      return +((valueNow - valueMin) / (valueMax - valueMin) * 100).toFixed(1);
    }

    return +((valueNow - valueMin) / (valueMax - valueMin + 1) * 100).toFixed(1);
  } // for every other handle stick on the right edge


  return +((valueNow - valueMin + 1) / (valueMax - valueMin + 1) * 100).toFixed(1);
};

exports.calculateLeftPosition = calculateLeftPosition;

var isFirst = function isFirst(value, valueNow, index, hasHistogram) {
  if (Array.isArray(value)) {
    var max = Math.max.apply(Math, _toConsumableArray(value));
    var min = Math.min.apply(Math, _toConsumableArray(value));
    var maxEqualsMin = max === min;
    var minEqualsValueNow = min === valueNow;
    if (index !== 0 && maxEqualsMin) return false;
    return maxEqualsMin || minEqualsValueNow;
  }

  return !hasHistogram;
};

exports.isFirst = isFirst;
var StyledHandle = (0, _styledComponents.default)(function (_ref) {
  var left = _ref.left,
      theme = _ref.theme,
      onTop = _ref.onTop,
      props = _objectWithoutProperties(_ref, ["left", "theme", "onTop"]);

  return React.createElement("div", props);
}).attrs(function (_ref2) {
  var _style;

  var left = _ref2.left,
      onTop = _ref2.onTop,
      theme = _ref2.theme;
  return {
    style: (_style = {}, _defineProperty(_style, (0, _rtl.left)({
      theme: theme
    }), "calc(".concat(left, "% - 12px)")), _defineProperty(_style, "zIndex", onTop ? 20 : undefined), _style)
  };
}).withConfig({
  displayName: "Handle__StyledHandle",
  componentId: "sc-1douguf-0"
})(["display:flex;position:absolute;right:0;bottom:0;flex:0 0 24px;align-items:center;justify-content:center;box-sizing:border-box;width:24px;height:24px;border-radius:24px;box-shadow:0 1px 4px 0 ", ";background-color:", ";cursor:pointer;transition:box-shadow ", " ease-in-out;-webkit-tap-highlight-color:transparent;-webkit-touch-callout:none;-webkit-user-select:none;:after{content:\"\";display:block;width:8px;height:8px;background-color:", ";border-radius:8px;}:hover{box-shadow:0 2px 6px 0 ", ";}:focus{outline:none;box-shadow:0 1px 4px 0 ", ",0 0 0 2px ", ";}:active{box-shadow:0 1px 4px 0 ", ",0 0 0 4px ", ";}"], function (_ref3) {
  var theme = _ref3.theme;
  return (0, _convertHexToRgba.default)(theme.orbit.paletteInkNormal, 10);
}, function (_ref4) {
  var theme = _ref4.theme;
  return theme.orbit.paletteWhite;
}, function (_ref5) {
  var theme = _ref5.theme;
  return theme.orbit.durationFast;
}, function (_ref6) {
  var theme = _ref6.theme;
  return theme.orbit.paletteProductNormal;
}, function (_ref7) {
  var theme = _ref7.theme;
  return (0, _convertHexToRgba.default)(theme.orbit.paletteInkNormal, 10);
}, function (_ref8) {
  var theme = _ref8.theme;
  return (0, _convertHexToRgba.default)(theme.orbit.paletteInkNormal, 10);
}, function (_ref9) {
  var theme = _ref9.theme;
  return (0, _convertHexToRgba.default)(theme.orbit.paletteProductNormal, 20);
}, function (_ref10) {
  var theme = _ref10.theme;
  return (0, _convertHexToRgba.default)(theme.orbit.paletteInkNormal, 10);
}, function (_ref11) {
  var theme = _ref11.theme;
  return (0, _convertHexToRgba.default)(theme.orbit.paletteProductNormal, 20);
});
StyledHandle.defaultProps = {
  theme: _defaultTheme.default
};

var Handle = function Handle(_ref12) {
  var tabIndex = _ref12.tabIndex,
      onMouseDown = _ref12.onMouseDown,
      onFocus = _ref12.onFocus,
      onTouchStart = _ref12.onTouchStart,
      valueMax = _ref12.valueMax,
      valueMin = _ref12.valueMin,
      onTop = _ref12.onTop,
      value = _ref12.value,
      index = _ref12.index,
      ariaValueText = _ref12.ariaValueText,
      ariaLabel = _ref12.ariaLabel,
      hasHistogram = _ref12.hasHistogram;
  var valueNow = Array.isArray(value) ? value[index] : value;
  var first = isFirst(value, valueNow, index, hasHistogram);
  var isSimple = !hasHistogram && !Array.isArray(value);
  var left = calculateLeftPosition(valueNow, valueMin, valueMax, first, isSimple);
  return React.createElement(StyledHandle, {
    tabIndex: tabIndex,
    onTop: onTop,
    role: "slider",
    onMouseDown: onMouseDown,
    onFocus: onFocus,
    onTouchStart: onTouchStart,
    "aria-valuemax": valueMax,
    "aria-valuemin": valueMin,
    "aria-valuenow": valueNow,
    "aria-label": Array.isArray(ariaLabel) && typeof index !== "undefined" ? ariaLabel[index] : ariaLabel,
    "aria-valuetext": ariaValueText,
    left: left
  });
};

var _default = Handle;
exports.default = _default;
