// @kiwicom/orbit-components/lib/Slider/components/Bar/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.StyledBarPart = exports.calculateBarPosition = void 0;

var React = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _rtl = require("../../../utils/rtl");

var _defaultTheme = _interopRequireDefault(require("../../../defaultTheme"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var StyledBar = _styledComponents.default.div.withConfig({
  displayName: "Bar__StyledBar",
  componentId: "rdurdt-0"
})(["display:block;position:relative;width:100%;height:24px;cursor:pointer;-webkit-tap-highlight-color:transparent;-webkit-touch-callout:none;-webkit-user-select:none;"]);

var calculateBarPosition = function calculateBarPosition(value, max, min, hasHistogram) {
  if (Array.isArray(value)) {
    return {
      left: +((value[0] - min) / (max - min + 1) * 100).toFixed(1),
      width: +((value[value.length - 1] - value[0] + 1) / (max - min + 1) * 100).toFixed(1)
    };
  }

  var addition = hasHistogram ? 1 : 0;
  return {
    left: 0,
    width: +((value - min + addition) / (max - min + addition) * 100).toFixed(1)
  };
};

exports.calculateBarPosition = calculateBarPosition;
var StyledBarPart = (0, _styledComponents.default)(function (_ref) {
  var width = _ref.width,
      left = _ref.left,
      theme = _ref.theme,
      active = _ref.active,
      props = _objectWithoutProperties(_ref, ["width", "left", "theme", "active"]);

  return React.createElement("div", props);
}).attrs(function (_ref2) {
  var width = _ref2.width,
      left = _ref2.left,
      theme = _ref2.theme;
  return {
    style: _defineProperty({
      width: "".concat(width, "%")
    }, (0, _rtl.left)({
      theme: theme
    }), "".concat(left, "%"))
  };
}).withConfig({
  displayName: "Bar__StyledBarPart",
  componentId: "rdurdt-1"
})(["position:absolute;height:4px;top:10px;border-radius:4px;background-color:", ";"], function (_ref3) {
  var theme = _ref3.theme,
      active = _ref3.active;
  return active ? theme.orbit.paletteProductNormal : theme.orbit.paletteInkLighter;
});
exports.StyledBarPart = StyledBarPart;
StyledBarPart.defaultProps = {
  theme: _defaultTheme.default
}; // $FlowExpected

var Bar = React.forwardRef(function (_ref4, ref) {
  var onMouseDown = _ref4.onMouseDown,
      value = _ref4.value,
      max = _ref4.max,
      min = _ref4.min,
      hasHistogram = _ref4.hasHistogram;

  var _calculateBarPosition = calculateBarPosition(value, max, min, hasHistogram),
      left = _calculateBarPosition.left,
      width = _calculateBarPosition.width;

  return React.createElement(StyledBar, {
    ref: ref,
    onMouseDown: onMouseDown
  }, React.createElement(StyledBarPart, {
    width: 100,
    left: 0
  }), React.createElement(StyledBarPart, {
    active: true,
    left: left,
    width: width
  }));
});
Bar.displayName = "Bar";
var _default = Bar;
exports.default = _default;
