// @kiwicom/orbit-components/lib/ServiceLogo/consts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.baseURL = exports.SIZE_OPTIONS = exports.NAME_OPTIONS = void 0;
var NAME_OPTIONS = {
  AIRHELP: "AirHelp",
  ALIPAY: "Alipay",
  AMEX: "Amex",
  AXA: "Axa",
  AXAASSISTANCE: "AxaAssistance",
  BUSINESSINSIDER: "BusinessInsider",
  CHINAUNIONPAY: "ChinaUnionPay",
  DAILYEXPRESS: "DailyExpress",
  DINERSCLUB: "DinersClub",
  DISCOVER: "Discover",
  IATA: "IATA",
  JCB: "JCB",
  MAILONLINE: "MailOnline",
  MAESTRO: "Maestro",
  MASTERCARD: "MasterCard",
  MIR: "MIR",
  MIRROR: "Mirror",
  NEWYORKTIMES: "NewYorkTimes",
  NORTOSECURED: "NortonSecured",
  PAYPAL: "PayPal",
  SOFORT: "Sofort",
  TRAVELPULSE: "TravelPulse",
  TRUSTLY: "Trustly",
  USATODAY: "UsaToday",
  VISA: "Visa",
  VISAHQ: "VisaHQ",
  ZOOZ: "Zooz"
};
exports.NAME_OPTIONS = NAME_OPTIONS;
var SIZE_OPTIONS = {
  SMALL: "small",
  MEDIUM: "medium",
  LARGE: "large"
};
exports.SIZE_OPTIONS = SIZE_OPTIONS;
var baseURL = "//images.kiwi.com";
exports.baseURL = baseURL;
