// @kiwicom/orbit-components/lib/CallOutBanner/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _convertHexToRgba = _interopRequireDefault(require("@kiwicom/orbit-design-tokens/lib/convertHexToRgba"));

var _mediaQuery = _interopRequireDefault(require("../utils/mediaQuery"));

var _defaultTheme = _interopRequireDefault(require("../defaultTheme"));

var _Heading = _interopRequireDefault(require("../Heading"));

var _Stack = _interopRequireDefault(require("../Stack"));

var _Text = _interopRequireDefault(require("../Text"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

var StyledCallOutBanner = _styledComponents.default.div.withConfig({
  displayName: "CallOutBanner__StyledCallOutBanner",
  componentId: "sc-1ml9ckq-0"
})(["display:flex;flex-direction:column;align-items:flex-start;justify-content:flex-start;background:", ";border-radius:", ";padding:", ";", ";", ";"], function (_ref) {
  var theme = _ref.theme;
  return theme.orbit.paletteWhite;
}, function (_ref2) {
  var theme = _ref2.theme;
  return theme.orbit.borderRadiusSmall;
}, function (_ref3) {
  var theme = _ref3.theme;
  return theme.orbit.spaceMedium;
}, function (_ref4) {
  var onClick = _ref4.onClick;
  return onClick ? (0, _styledComponents.css)(["box-shadow:0 0 2px 0 ", ",0 1px 4px 0 ", ";transition:box-shadow ", " ease-in-out;cursor:pointer;:focus,:active,:hover{box-shadow:0 1px 4px 0 ", ",0 4px 8px 0 ", ";outline:none;}border:1px solid transparent;"], function (_ref5) {
    var theme = _ref5.theme;
    return (0, _convertHexToRgba.default)(theme.orbit.paletteInkNormal, 16);
  }, function (_ref6) {
    var theme = _ref6.theme;
    return (0, _convertHexToRgba.default)(theme.orbit.paletteInkNormal, 12);
  }, function (_ref7) {
    var theme = _ref7.theme;
    return theme.orbit.durationFast;
  }, function (_ref8) {
    var theme = _ref8.theme;
    return (0, _convertHexToRgba.default)(theme.orbit.paletteInkNormal, 16);
  }, function (_ref9) {
    var theme = _ref9.theme;
    return (0, _convertHexToRgba.default)(theme.orbit.paletteInkNormal, 12);
  }) : (0, _styledComponents.css)(["border:1px solid rgb(232,237,241);"]);
}, _mediaQuery.default.largeMobile((0, _styledComponents.css)(["flex-direction:row;padding:", ";"], function (_ref10) {
  var theme = _ref10.theme;
  return theme.orbit.spaceLarge;
})));

StyledCallOutBanner.defaultProps = {
  theme: _defaultTheme.default
};

var StyledIllustration = _styledComponents.default.div.withConfig({
  displayName: "CallOutBanner__StyledIllustration",
  componentId: "sc-1ml9ckq-1"
})(["padding-bottom:", ";", ";"], function (_ref11) {
  var theme = _ref11.theme;
  return theme.orbit.spaceMedium;
}, _mediaQuery.default.largeMobile((0, _styledComponents.css)(["padding-right:", ";padding-bottom:0;"], function (_ref12) {
  var theme = _ref12.theme;
  return theme.orbit.spaceLarge;
})));

StyledIllustration.defaultProps = {
  theme: _defaultTheme.default
};

var CallOutBanner = function CallOutBanner(_ref13) {
  var children = _ref13.children,
      illustration = _ref13.illustration,
      actions = _ref13.actions,
      title = _ref13.title,
      onClick = _ref13.onClick,
      tabIndex = _ref13.tabIndex,
      description = _ref13.description,
      dataTest = _ref13.dataTest;
  return React.createElement(StyledCallOutBanner, {
    onClick: onClick,
    tabIndex: (onClick || tabIndex) && (tabIndex || 0),
    "data-test": dataTest
  }, illustration && React.createElement(StyledIllustration, null, illustration), React.createElement(_Stack.default, {
    spacing: "natural"
  }, React.createElement(_Stack.default, {
    spacing: "condensed"
  }, React.createElement(_Stack.default, {
    spacing: "tight"
  }, title && React.createElement(_Heading.default, {
    type: "title3"
  }, title), description && React.createElement(_Text.default, null, description)), children), actions));
};

var _default = CallOutBanner;
exports.default = _default;
