// @kiwicom/orbit-components/lib/Stepper/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "StepperStateless", {
  enumerable: true,
  get: function get() {
    return _StepperStateless.default;
  }
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _StepperStateless = _interopRequireDefault(require("./StepperStateless"));

var _validateIncrement = _interopRequireDefault(require("../utils/validateIncrement"));

var _validateDecrement = _interopRequireDefault(require("../utils/validateDecrement"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Stepper =
/*#__PURE__*/
function (_React$PureComponent) {
  _inherits(Stepper, _React$PureComponent);

  function Stepper() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, Stepper);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(Stepper)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "state", {
      value: _this.props.defaultValue || 0
    });

    _defineProperty(_assertThisInitialized(_this), "setValueAndInjectCallback", function (value) {
      var onChange = _this.props.onChange;

      if (onChange) {
        onChange(value);
      }

      _this.setState({
        value: value
      });
    });

    _defineProperty(_assertThisInitialized(_this), "incrementCounter", function () {
      var value = _this.state.value;
      var _this$props = _this.props,
          _this$props$maxValue = _this$props.maxValue,
          maxValue = _this$props$maxValue === void 0 ? Number.POSITIVE_INFINITY : _this$props$maxValue,
          _this$props$step = _this$props.step,
          step = _this$props$step === void 0 ? 1 : _this$props$step;

      _this.setValueAndInjectCallback((0, _validateIncrement.default)({
        value: value,
        maxValue: maxValue,
        step: step
      }));
    });

    _defineProperty(_assertThisInitialized(_this), "decrementCounter", function () {
      var value = _this.state.value;
      var _this$props2 = _this.props,
          _this$props2$minValue = _this$props2.minValue,
          minValue = _this$props2$minValue === void 0 ? Number.NEGATIVE_INFINITY : _this$props2$minValue,
          _this$props2$step = _this$props2.step,
          step = _this$props2$step === void 0 ? 1 : _this$props2$step;

      _this.setValueAndInjectCallback((0, _validateDecrement.default)({
        value: value,
        minValue: minValue,
        step: step
      }));
    });

    _defineProperty(_assertThisInitialized(_this), "handleKeyDown", function (ev) {
      if (ev.keyCode === 40) {
        ev.preventDefault();

        _this.decrementCounter();
      }

      if (ev.keyCode === 38) {
        ev.preventDefault();

        _this.incrementCounter();
      }
    });

    return _this;
  }

  _createClass(Stepper, [{
    key: "render",
    value: function render() {
      var _this$props3 = this.props,
          onBlur = _this$props3.onBlur,
          onFocus = _this$props3.onFocus,
          disabled = _this$props3.disabled,
          name = _this$props3.name,
          dataTest = _this$props3.dataTest,
          maxValue = _this$props3.maxValue,
          minValue = _this$props3.minValue,
          titleIncrement = _this$props3.titleIncrement,
          titleDecrement = _this$props3.titleDecrement;
      var value = this.state.value;
      return React.createElement(_StepperStateless.default, {
        disabled: disabled,
        dataTest: dataTest,
        value: value,
        name: name,
        minValue: minValue,
        maxValue: maxValue,
        onKeyDown: this.handleKeyDown,
        onBlur: onBlur,
        onFocus: onFocus,
        onIncrement: this.incrementCounter,
        onDecrement: this.decrementCounter,
        titleIncrement: titleIncrement,
        titleDecrement: titleDecrement
      });
    }
  }]);

  return Stepper;
}(React.PureComponent);

var _default = Stepper;
exports.default = _default;
