// @kiwicom/orbit-components/lib/PricingTable/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "PricingTableItem", {
  enumerable: true,
  get: function get() {
    return _PricingTableItem.default;
  }
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Stack = _interopRequireDefault(require("../Stack"));

var _Text = _interopRequireDefault(require("../Text"));

var _useMediaQuery2 = _interopRequireDefault(require("../hooks/useMediaQuery"));

var _PricingTableItem = _interopRequireDefault(require("./PricingTableItem"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var StyledPricingTable = _styledComponents.default.div.withConfig({
  displayName: "PricingTable__StyledPricingTable",
  componentId: "sc-1aibxyh-0"
})([""]);

var PricingTable = function PricingTable(_ref) {
  var children = _ref.children,
      _ref$defaultActiveEle = _ref.defaultActiveElement,
      defaultActiveElement = _ref$defaultActiveEle === void 0 ? 0 : _ref$defaultActiveEle;

  var _useMediaQuery = (0, _useMediaQuery2.default)(),
      isTablet = _useMediaQuery.isTablet;

  var _useState = (0, _react.useState)(defaultActiveElement),
      _useState2 = _slicedToArray(_useState, 2),
      activeElement = _useState2[0],
      setActiveElement = _useState2[1];

  var handleOnClick = function handleOnClick(i) {
    return function () {
      if (!isTablet) {
        setActiveElement(i);
      }
    };
  };

  var resolveBasis = function resolveBasis(item) {
    if (item.length) {
      return "".concat(Math.floor(100 / item.length), "%");
    }

    return "100%";
  };

  return _react.default.createElement(StyledPricingTable, null, _react.default.createElement(_Stack.default, {
    flex: true,
    grow: true,
    spaceAfter: "medium",
    spacing: "condensed",
    align: "stretch",
    tablet: {
      spacing: "natural",
      spaceAfter: "none"
    }
  }, isTablet ? children : _react.default.Children.map(children, function (child, i) {
    return _react.default.cloneElement(child, {
      active: activeElement === i,
      compact: true,
      basis: resolveBasis(child),
      onClick: handleOnClick(i)
    });
  })), !isTablet && children && _react.default.createElement(_Stack.default, {
    spacing: "condensed"
  }, _react.default.Children.map(children, function (child, i) {
    if (i === activeElement) {
      return _react.default.createElement(_react.default.Fragment, null, child.props.mobileDescription && _react.default.createElement(_Text.default, {
        weight: "bold",
        size: "normal"
      }, child.props.mobileDescription), child.props.children && _react.default.cloneElement(child.props.children), child.props.action && _react.default.cloneElement(child.props.action));
    }

    return null;
  })));
};

var _default = PricingTable;
exports.default = _default;
