// @kiwicom/orbit-components/lib/Pagination/components/ActiveButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _Button = _interopRequireDefault(require("../../Button"));

var _defaultTheme = _interopRequireDefault(require("../../defaultTheme"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var StyledButton = _styledComponents.default.div.withConfig({
  displayName: "ActiveButton__StyledButton",
  componentId: "sc-1j3a7nx-0"
})(["&:hover,&:active,&:focus{background:", ";color:", ";transform:none;cursor:default;}", ";"], function (_ref) {
  var theme = _ref.theme;
  return theme.orbit.backgroundButtonSecondary;
}, function (_ref2) {
  var theme = _ref2.theme;
  return theme.orbit.colorTextButtonSecondary;
}, function (_ref3) {
  var transparent = _ref3.transparent;
  return transparent && (0, _styledComponents.css)(["&&&{background-color:transparent;}"]);
});

StyledButton.defaultProps = {
  theme: _defaultTheme.default
};

var ActiveButton = function ActiveButton(_ref4) {
  var children = _ref4.children,
      transparent = _ref4.transparent,
      size = _ref4.size;
  return React.createElement(_Button.default, {
    type: "secondary",
    size: size,
    component: function component(props) {
      return React.createElement(StyledButton, _extends({}, props, {
        type: undefined,
        transparent: transparent
      }));
    }
  }, children);
};

var _default = ActiveButton;
exports.default = _default;
