// @kiwicom/orbit-components/lib/Stepper/StepperStateless/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Button = _interopRequireDefault(require("../../Button"));

var _Minus = _interopRequireDefault(require("../../icons/Minus"));

var _Plus = _interopRequireDefault(require("../../icons/Plus"));

var _defaultTheme = _interopRequireDefault(require("../../defaultTheme"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

var StyledStepper = _styledComponents.default.div.withConfig({
  displayName: "StepperStateless__StyledStepper",
  componentId: "dh3icj-0"
})(["display:flex;width:100%;flex:1 1 auto;"]);

var StyledStepperInput = _styledComponents.default.input.withConfig({
  displayName: "StepperStateless__StyledStepperInput",
  componentId: "dh3icj-1"
})(["width:100%;height:32px;padding:0;border:0;font-size:", ";font-weight:", ";color:", ";text-align:center;min-width:0;&::-webkit-inner-spin-button,&::-webkit-outer-spin-button{-webkit-appearance:none;margin:0;}&:focus{outline:none;}"], function (_ref) {
  var theme = _ref.theme;
  return theme.orbit.fontSizeInputNormal;
}, function (_ref2) {
  var theme = _ref2.theme;
  return theme.orbit.fontWeightBold;
}, function (_ref3) {
  var theme = _ref3.theme;
  return theme.orbit.paletteInkNormal;
});

StyledStepperInput.defaultProps = {
  theme: _defaultTheme.default
};

var StepperStateless = function StepperStateless(_ref4) {
  var disabled = _ref4.disabled,
      dataTest = _ref4.dataTest,
      value = _ref4.value,
      name = _ref4.name,
      minValue = _ref4.minValue,
      maxValue = _ref4.maxValue,
      _onKeyDown = _ref4.onKeyDown,
      onBlur = _ref4.onBlur,
      onFocus = _ref4.onFocus,
      onIncrement = _ref4.onIncrement,
      onDecrement = _ref4.onDecrement,
      titleIncrement = _ref4.titleIncrement,
      titleDecrement = _ref4.titleDecrement,
      disabledIncrement = _ref4.disabledIncrement,
      disabledDecrement = _ref4.disabledDecrement;
  return React.createElement(StyledStepper, {
    "data-test": dataTest
  }, React.createElement(_Button.default, {
    disabled: disabled || disabledDecrement || typeof value === "number" && value <= +minValue,
    iconLeft: React.createElement(_Minus.default, null),
    type: "secondary",
    size: "small",
    onClick: function onClick(ev) {
      if (onDecrement) {
        onDecrement(ev);
      }
    },
    title: titleDecrement
  }), React.createElement(StyledStepperInput, {
    name: name,
    disabled: disabled,
    type: "text",
    value: value || 0,
    min: minValue,
    max: maxValue,
    onKeyDown: function onKeyDown(ev) {
      if (_onKeyDown) {
        _onKeyDown(ev);
      }
    },
    onBlur: onBlur,
    onFocus: onFocus,
    readOnly: true
  }), React.createElement(_Button.default, {
    disabled: disabled || disabledIncrement || typeof value === "number" && value >= +maxValue,
    iconLeft: React.createElement(_Plus.default, null),
    type: "secondary",
    size: "small",
    onClick: function onClick(ev) {
      if (onIncrement) {
        onIncrement(ev);
      }
    },
    title: titleIncrement
  }));
};

var _default = StepperStateless;
exports.default = _default;
