// @kiwicom/orbit-components/lib/InputStepper/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "InputStepperStateless", {
  enumerable: true,
  get: function get() {
    return _InputStepperStateless.default;
  }
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _consts = require("../InputField/consts");

var _keyMaps = _interopRequireDefault(require("../common/keyMaps"));

var _InputStepperStateless = _interopRequireDefault(require("./InputStepperStateless"));

var _validateIncrement = _interopRequireDefault(require("../utils/validateIncrement"));

var _validateDecrement = _interopRequireDefault(require("../utils/validateDecrement"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var InputStepper =
/*#__PURE__*/
function (_React$Component) {
  _inherits(InputStepper, _React$Component);

  function InputStepper() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, InputStepper);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(InputStepper)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "state", {
      value: _this.props.defaultValue || 0
    });

    _defineProperty(_assertThisInitialized(_this), "setValueAndInjectCallback", function (value) {
      var onChange = _this.props.onChange;

      if (onChange) {
        onChange(value);
      }

      _this.setState({
        value: value
      });
    });

    _defineProperty(_assertThisInitialized(_this), "incrementCounter", function () {
      var value = _this.state.value;
      var _this$props = _this.props,
          _this$props$maxValue = _this$props.maxValue,
          maxValue = _this$props$maxValue === void 0 ? Number.POSITIVE_INFINITY : _this$props$maxValue,
          _this$props$step = _this$props.step,
          step = _this$props$step === void 0 ? 1 : _this$props$step;

      _this.setValueAndInjectCallback((0, _validateIncrement.default)({
        value: value,
        maxValue: maxValue,
        step: step
      }));
    });

    _defineProperty(_assertThisInitialized(_this), "decrementCounter", function () {
      var value = _this.state.value;
      var _this$props2 = _this.props,
          _this$props2$minValue = _this$props2.minValue,
          minValue = _this$props2$minValue === void 0 ? Number.NEGATIVE_INFINITY : _this$props2$minValue,
          _this$props2$step = _this$props2.step,
          step = _this$props2$step === void 0 ? 1 : _this$props2$step;

      _this.setValueAndInjectCallback((0, _validateDecrement.default)({
        value: value,
        minValue: minValue,
        step: step
      }));
    });

    _defineProperty(_assertThisInitialized(_this), "handleIncrementCounter", function (ev) {
      if (ev && ev.type === "click") {
        _this.incrementCounter();
      }

      if (ev && ev.type === "keydown") {
        if (ev.keyCode === _keyMaps.default.SPACE) {
          ev.preventDefault();

          _this.incrementCounter();
        } else if (ev.keyCode === _keyMaps.default.ENTER) {
          _this.incrementCounter();
        }
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleDecrementCounter", function (ev) {
      if (ev && ev.type === "click") {
        _this.decrementCounter();
      }

      if (ev && ev.type === "keydown") {
        if (ev.keyCode === _keyMaps.default.SPACE) {
          ev.preventDefault();

          _this.decrementCounter();
        } else if (ev.keyCode === _keyMaps.default.ENTER) {
          _this.decrementCounter();
        }
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleKeyDown", function (ev) {
      if (ev.keyCode === _keyMaps.default.ARROW_DOWN) {
        ev.preventDefault();

        _this.decrementCounter();
      }

      if (ev.keyCode === _keyMaps.default.ARROW_UP) {
        ev.preventDefault();

        _this.incrementCounter();
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleChange", function (ev) {
      var _this$props3 = _this.props,
          _this$props3$minValue = _this$props3.minValue,
          minValue = _this$props3$minValue === void 0 ? Number.NEGATIVE_INFINITY : _this$props3$minValue,
          _this$props3$maxValue = _this$props3.maxValue,
          maxValue = _this$props3$maxValue === void 0 ? Number.POSITIVE_INFINITY : _this$props3$maxValue;
      var value = ev && parseInt(ev.target.value, 10);
      var prevValue = _this.state.value;

      if (prevValue <= value) {
        _this.setState({
          value: (0, _validateIncrement.default)({
            value: value,
            maxValue: maxValue,
            step: 0
          })
        });
      }

      if (prevValue >= value) {
        _this.setState({
          value: (0, _validateDecrement.default)({
            value: value,
            minValue: minValue,
            step: 0
          })
        });
      }
    });

    return _this;
  }

  _createClass(InputStepper, [{
    key: "render",
    value: function render() {
      var _this$props4 = this.props,
          label = _this$props4.label,
          error = _this$props4.error,
          onBlur = _this$props4.onBlur,
          onFocus = _this$props4.onFocus,
          help = _this$props4.help,
          disabled = _this$props4.disabled,
          name = _this$props4.name,
          dataTest = _this$props4.dataTest,
          _this$props4$size = _this$props4.size,
          size = _this$props4$size === void 0 ? _consts.SIZE_OPTIONS.NORMAL : _this$props4$size,
          maxValue = _this$props4.maxValue,
          minValue = _this$props4.minValue,
          required = _this$props4.required,
          tabIndex = _this$props4.tabIndex,
          forwardedRef = _this$props4.forwardedRef,
          spaceAfter = _this$props4.spaceAfter,
          titleIncrement = _this$props4.titleIncrement,
          titleDecrement = _this$props4.titleDecrement;
      var value = this.state.value;
      return React.createElement(_InputStepperStateless.default, {
        dataTest: dataTest,
        size: size,
        label: label,
        disabled: disabled,
        required: required,
        name: name,
        error: error,
        help: help,
        onChange: this.handleChange,
        onBlur: onBlur,
        onFocus: onFocus,
        onKeyDown: this.handleKeyDown,
        value: value || 0,
        minValue: minValue,
        maxValue: maxValue,
        tabIndex: tabIndex,
        forwardedRef: forwardedRef,
        spaceAfter: spaceAfter,
        onDecrement: this.handleDecrementCounter,
        onIncrement: this.handleIncrementCounter,
        titleIncrement: titleIncrement,
        titleDecrement: titleDecrement
      });
    }
  }]);

  return InputStepper;
}(React.Component); // $FlowExpected


var ForwardedInputStepper = React.forwardRef(function (props, ref) {
  return React.createElement(InputStepper, _extends({
    forwardedRef: ref
  }, props));
});
ForwardedInputStepper.displayName = "InputStepper";
var _default = ForwardedInputStepper;
exports.default = _default;
